rust/datafusion/src/execution/physical_plan/string_expressions.rs

```diff
@@ -0,0 +1,63 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! String expressions
+
+use crate::error::{ExecutionError, Result};
+use arrow::array::{Array, ArrayRef, StringArray, StringBuilder};
+
+macro_rules! downcast_vec {
+    ($ARGS:expr, $ARRAY_TYPE:ident) => {{
+        $ARGS
+            .iter()
+            .map(|e| match e.as_any().downcast_ref::<$ARRAY_TYPE>() {
+                Some(array) => Ok(array),
+                _ => Err(ExecutionError::General("failed to downcast".to_string())),
+            })
+    }};
+}
+
+/// concatenate string columns together.
+pub fn concatenate(args: &[ArrayRef]) -> Result<StringArray> {
+    // downcast all arguments to strings
+    let args = downcast_vec!(args, StringArray).collect::<Result<Vec<&StringArray>>>()?;
+    // do not accept 0 arguments.
+    assert!(args.len() != 0);
```

Comment:
this currently panics on zero arguments, which means that the plan `SELECT c1, concat() FROM t` crashes on run-time.

We should prob. safe-guard this in type coercion.

Essentially, all our physical expressions currently require one argument or they do not know how many rows to return.

Reply:
I would prefer that we not commit code that can `panic`s -- so in this PR I would suggest doing one of the following
1. Add an error in physical planing for `concat()`
2. Implement `concat()`

Here is one possible behavior for `concat()`:
1. The result value is always the Null
2. It should produce one output value for each input value

A potential implementation could be "rewrite `concat()` to be `concat("")` -- aka have the planner insert a single constant argument

Reply:
Thanks for the suggestions 👍 I've added an error when the plan has 0 arguments.